### ISSUE TITLE
Belos: potential issue in block CG

### ISSUE THREAD

**Issue by jhux2** [trilinos/Trilinos#134]
A user has reported a potential issue in Belos' CG method.  This is documented at UK-MAC/TeaLeaf_Trilinos/issues/2.

@trilinos/belos

**Comment by jhux2** [comment]
Comment from @rsmedleystevenson

>    As I mentioned last week I have discovered an issue with one of the Belos block Krylov solvers, which is described in the link included below. This relates to left preconditioning of the PseudoBlockCG algorithm in Belos for a repeated solve of a time-evolution problem in TeaLeaf.
>
>   I've managed to come up with a fix but am not sure of the exact cause of the problem (due to my lack of familiarity of the package and C++ in general). However, I think it is due to a logical inconsistency in the implementation itself rather than a more subtle bug.
>
>   Please feel free to pass on this message to the relevant developer. Hopefully they can reproduce the fault.  Details of my compilation environment are included for reference:


```
rsmedley@anca:~/TeaLeaf_Trilinos$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.8/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 4.8.4-2ubuntu1~14.04' --with-bugurl=file:///usr/share/doc/gcc-4.8/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.8 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.8 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libmudflap --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.8-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.8-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu

Thread model: posix
gcc version 4.8.4 (Ubuntu 4.8.4-2ubuntu1~14.04) ```

**Comment by mhoemmen** [comment]
@jhux2 Do the TeaLeaf folks have a patch?

**Comment by jhux2** [comment]
> Do the TeaLeaf folks have a patch?

That was my impression, but I haven't seen it myself.

**Comment by mhoemmen** [comment]
I posted a comment on their Issue asking if they had a patch.  Would rather use theirs if they know it works :-)

**Comment by rsmedleystevenson** [comment]
I've added an additional comment at https://github.com/UK-MAC/TeaLeaf_Trilinos/issues/2 to clarify the fix, which was to force a recalculation of the residual.

**Comment by rsmedleystevenson** [comment]
I did not manage to track down why the residual value was incorrect i.e. it was set to the value of z=M^{-1}r rather than r=b-Ax when the left preconditioner was applied. Then the left preconditioner was applied again to r leading to z -> M^{-1}M^{-1}(b-Ax), which is wrong. The fix just reset r to be correct.

**Comment by hkthorn** [comment]
When a Belos solver object and linear problem object are being reused for multiple solves, then the solver should be "reset". That communicates to the solver that another solve is happening and let's them do whatever reinitialization that may be necessary. So, in between calls to "solve()" (after you changed the linear system values) did you call "reset( Belos::Problem )" to communicate to the solver that the RHS vector and matrix entry values had changed? That would prompt the solver to compute the correct initial residual from the changed entries in the linear problem.

There is an example of different types of solver object reuse in the test: belos/epetra/test/BlockCG/test_resolve_cg_hb.cpp

Maybe that is the issue?

Thanks,
Heidi

--

Heidi K. Thornquist
Electrical Models & Simulation
Sandia National Laboratories
Albuquerque, NM  87185-1323

**Comment by hkthorn** [comment]
FYI, for those interested in the resolution to this issue.  We have worked through the logic in Belos to find the cause:

"I was calling setProblem on the LinearProblem object prior to setting the preconditioner and I think this was the cause of the error. Moving this to after setting the preconditioner makes the issue go away.  I think this is the true source of the issue, which comes down to a mixture of user error :-) and differences between the internal workings of the PseudoBlockCG and RCG solver.

Heidi, I'd suggest you close the Trilinos ticket assuming you're happy with this explanation, unless you have a suggestion for how the code could be modified to prevent this re-occurring?"

I have added a small piece of logic to the setProblem method of the LinearProblem class that makes sure that the preconditioned residual vector pointer is not pointing to the residual vector pointer if there is a left-preconditioner, which is the source of this issue.

Thanks,

Heidi

--

Heidi K. Thornquist
Electrical Models & Simulation
Sandia National Laboratories
Albuquerque, NM  87185-1323
